### PR TITLE
ENH: DataFrame.pivot accepts a list of values

### PR DESCRIPTION
**Pull request by ibrahimsharaf**
- [x] closes https://github.com/pandas-dev/pandas/issues/17160
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/reshape/test_pivot.py [hunk 1]
```diff
@@ -353,6 +353,29 @@ def test_pivot_periods(self):
         pv = df.pivot(index='p1', columns='p2', values='data1')
         tm.assert_frame_equal(pv, expected)
 
+    def test_pivot_with_multi_values(self):
+        df = pd.DataFrame({'foo': ['one', 'one', 'one', 'two', 'two', 'two'],
+                           'bar': ['A', 'B', 'C', 'A', 'B', 'C'],
+                           'baz': [1, 2, 3, 4, 5, 6],
+                           'zoo': ['x', 'y', 'z', 'q', 'w', 't']})
+
+        results = df.pivot(index='zoo', columns='foo', values=['bar', 'baz'])
```
Comment (jreback) [inline pandas/tests/reshape/test_pivot.py]: use result=

File: pandas/tests/reshape/test_pivot.py [hunk 2]
```diff
@@ -353,6 +353,29 @@ def test_pivot_periods(self):
...
+
+        data = [[None, 'A', None, 4],
```
Comment (jreback) [inline pandas/tests/reshape/test_pivot.py]: use np.nan rather than None

File: pandas/tests/reshape/test_pivot.py [hunk 3]
```diff
@@ -353,6 +353,29 @@ def test_pivot_periods(self):
         pv = df.pivot(index='p1', columns='p2', values='data1')
         tm.assert_frame_equal(pv, expected)
 
+    def test_pivot_with_multi_values(self):
+        df = pd.DataFrame({'foo': ['one', 'one', 'one', 'two', 'two', 'two'],
```
Comment (jreback) [inline pandas/tests/reshape/test_pivot.py]: add the issue number as a comment

File: pandas/tests/reshape/test_pivot.py [hunk 4]
```diff
@@ -353,6 +353,29 @@ def test_pivot_periods(self):
         pv = df.pivot(index='p1', columns='p2', values='data1')
         tm.assert_frame_equal(pv, expected)
 
+    def test_pivot_with_multi_values(self):
```
Comment (jreback) [inline pandas/tests/reshape/test_pivot.py]: say with list_like_values rather than multi_values

File: doc/source/whatsnew/v0.22.0.txt [hunk 5]
```diff
@@ -129,6 +124,10 @@ Other API Changes
 - :func:`DataFrame.from_items` provides a more informative error message when passed scalar values (:issue:`17312`)
 - When created with duplicate labels, ``MultiIndex`` now raises a ``ValueError``. (:issue:`17464`)
 - Building from source now explicity requires ``setuptools`` in ``setup.py`` (:issue:`18113`)
+- :func:`Series.fillna` now raises a ``TypeError`` instead of a ``ValueError`` when passed a list, tuple or DataFrame as a ``value`` (:issue:`18293`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.22.0.txt]: looks like you picked up some other commits here.

File: pandas/tests/reshape/test_pivot.py [hunk 6]
```diff
@@ -353,6 +353,29 @@ def test_pivot_periods(self):
...
+                           'baz': [1, 2, 3, 4, 5, 6],
+                           'zoo': ['x', 'y', 'z', 'q', 'w', 't']})
+
+        results = df.pivot(index='zoo', columns='foo', values=['bar', 'baz'])
+
+        data = [[None, 'A', None, 4],
+                [None, 'C', None, 6],
+                [None, 'B', None, 5],
+                ['A', None, 1, None],
+                ['B', None, 2, None],
```
Comment (jreback) [inline pandas/tests/reshape/test_pivot.py]: parmaterize this with values a list, tuple, np.array, pd.Series, pd.Index (should all act the same)

File: pandas/core/reshape/reshape.py [hunk 7]
```diff
@@ -374,15 +374,17 @@ def pivot(self, index=None, columns=None, values=None):
         cols = [columns] if index is None else [index, columns]
         append = index is None
         indexed = self.set_index(cols, append=append)
-        return indexed.unstack(columns)
     else:
-        if index is None:
-            index = self.index
+        index = self.index if index is None else self[index]
+        index = MultiIndex.from_arrays([index, self[columns]])
+        if isinstance(values, list):
```
Comment (jreback) [inline pandas/core/reshape/reshape.py]: use`` is_list_like`` here

Comment (jreback) [reply pandas/core/reshape/reshape.py]: add a comment here on what is going on

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.22.0.txt (hunk 5 above)
Comment (ibrahimsharaf) [reply doc/source/whatsnew/v0.22.0.txt]: those are merged on master now, so I think merging them as is won't harm (or duplicate).

File: pandas/tests/reshape/test_pivot.py (hunk 1 above)
Comment (ibrahimsharaf) [reply pandas/tests/reshape/test_pivot.py]: Can you elaborate why?

**Comment by ibrahimsharaf** [comment]
@jreback should the [documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.pivot.html) be modified at this PR?

File: doc/source/whatsnew/v0.22.0.txt (hunk 5 above)
Comment (jorisvandenbossche) [reply doc/source/whatsnew/v0.22.0.txt]: I am not sure if it will do harm or not, but can you still fix this? (it makes reviewing harder as there are unrelated changes) 
In principle

```
git fetch upstream
git merge upstream/master
```

should do it

**Comment by jorisvandenbossche** [comment]
> @jreback should the documentation page be modified at this PR?

Yes, you need to update the docstring / tutorial docs in this PR.

Can you show an actual example of what this PR can do? (eg show the result of the df in the test) As the test example result looks a bit strange from a quick look (all the NaNs)

File: doc/source/whatsnew/v0.22.0.txt (hunk 5 above)
Comment (ibrahimsharaf) [reply doc/source/whatsnew/v0.22.0.txt]: Merged

**Comment by ibrahimsharaf** [comment]
@jreback can you help me find the reason of the failing test? it happens when I pass the values columns as a tuple in those lines ([test](https://github.com/pandas-dev/pandas/pull/18636/files#diff-52cd4a8cbeafc8b5ec9110e1de9d023fR384), [def](https://github.com/pandas-dev/pandas/pull/18636/files#diff-fef81b7e498e469973b2da18d19ff6f3R382))

I tried adding tuple type in this condition ([core/frame.py](https://github.com/pandas-dev/pandas/blob/master/pandas/core/frame.py#L2150)), it passed this test, but failed many others.

**Comment by jorisvandenbossche** [comment]
The reason of the error is that a tuple `('bar', 'baz')` is seen as a single column name (for a MultiIndex), which is the correct behaviour IMO. So you shouldn't try to interpret a list the same as a tuple.

**Comment by ibrahimsharaf** [comment]
So what you are saying is that dealing with a tuple of values should be split from the `is_list_like` condition in pivot function?

**Comment by ibrahimsharaf** [comment]
@jreback @jorisvandenbossche would you guys take a look at the changes?

File: doc/source/whatsnew/v0.22.0.txt [hunk 8]
```diff
@@ -139,6 +139,8 @@ Other Enhancements
 - :func:`read_excel()` has gained the ``nrows`` parameter (:issue:`16645`)
 - :func:``DataFrame.to_json`` and ``Series.to_json`` now accept an ``index`` argument which allows the user to exclude the index from the JSON output (:issue:`17394`)
 - ``IntervalIndex.to_tuples()`` has gained the ``na_tuple`` parameter to control whether NA is returned as a tuple of NA, or NA itself (:issue:`18756`)
+- :func:`DataFrame.pivot` now accepts a list of values (:issue:`17160`).
```
Comment (jreback) [inline doc/source/whatsnew/v0.22.0.txt]: add for the ``values=`` kwargs

File: pandas/core/frame.py [hunk 9]
```diff
@@ -4390,6 +4391,19 @@ def pivot(self, index=None, columns=None, values=None):
         one  1   2   3
         two  4   5   6
 
+       >>> df.pivot(index='foo', columns='bar', values=['baz', 'zoo'])
```
Comment (jreback) [inline pandas/core/frame.py]: can you show an example using a single value first

File: pandas/core/reshape/reshape.py [hunk 10]
```diff
@@ -374,15 +374,18 @@ def pivot(self, index=None, columns=None, values=None):
         cols = [columns] if index is None else [index, columns]
         append = index is None
         indexed = self.set_index(cols, append=append)
-        return indexed.unstack(columns)
     else:
-        if index is None:
-            index = self.index
+        index = self.index if index is None else self[index]
+        index = MultiIndex.from_arrays([index, self[columns]])
+        if is_list_like(values):
```
Comment (jreback) [inline pandas/core/reshape/reshape.py]: I think you need:

``is_list_like(values) and not isinstance(values, tuple)``

File: pandas/tests/reshape/test_pivot.py [hunk 11]
```diff
@@ -371,6 +371,40 @@ def test_pivot_periods(self):
...
         tm.assert_frame_equal(pv, expected)
 
+    def test_pivot_with_list_like_values(self):
+        # issue #17160
+        df = pd.DataFrame({'foo': ['one', 'one', 'one', 'two', 'two', 'two'],
+                           'bar': ['A', 'B', 'C', 'A', 'B', 'C'],
+                           'baz': [1, 2, 3, 4, 5, 6],
+                           'zoo': ['x', 'y', 'z', 'q', 'w', 't']})
+
+        result_list = df.pivot(index='zoo', columns='foo',
```
Comment (jreback) [inline pandas/tests/reshape/test_pivot.py]: pls parametrize on these types, somethign like

```
@pytest.mark.parametrize("constructor", [pd.Series, pd.idnex, np.array, list]))
```

File: pandas/tests/reshape/test_pivot.py [hunk 12]
```diff
@@ -371,6 +371,40 @@ def test_pivot_periods(self):
...
+                                values=pd.Index(['bar', 'baz']))
+
+        data = [[np.nan, 'A', np.nan, 4],
+                [np.nan, 'C', np.nan, 6],
+                [np.nan, 'B', np.nan, 5],
+                ['A', np.nan, 1, np.nan],
+                ['B', np.nan, 2, np.nan],
+                ['C', np.nan, 3, np.nan]]
+        index = Index(data=['q', 't', 'w', 'x', 'y', 'z'], name='zoo')
+        columns = MultiIndex(levels=[['bar', 'baz'], ['one', 'two']],
```
Comment (jreback) [inline pandas/tests/reshape/test_pivot.py]: add a test with a MultiIndex and pass values as a tuple

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/reshape/test_pivot.py (hunk 12 above)
Comment (ibrahimsharaf) [reply pandas/tests/reshape/test_pivot.py]: Are you looking for something like the following? (please correct me if I am wrong)
```
    bar          baz       
  first second first second
0   one      A     1      x
1   one      B     2      y
2   one      C     3      z
3   two      A     4      q
4   two      B     5      w
5   two      C     6      t 
```

then pivot it:
``` 
df.pivot(index=('bar', 'first'), columns=('bar', 'second'), values=('baz', 'first'))
 ```

so the output would be:
```
     A   B   C
one  1   2   3
two  4   5   6
```

File: doc/source/whatsnew/v0.22.0.txt (hunk 8 above)
Comment (ibrahimsharaf) [reply doc/source/whatsnew/v0.22.0.txt]: I'm not sure I get you

File: pandas/tests/reshape/test_pivot.py (hunk 12 above)
Comment (ibrahimsharaf) [reply pandas/tests/reshape/test_pivot.py]: ping @jreback

File: doc/source/whatsnew/v0.22.0.txt (hunk 8 above)
Comment (jreback) [reply doc/source/whatsnew/v0.22.0.txt]: now accepts a list for the ``values=`` kwarg.

File: pandas/core/frame.py [hunk 13]
```diff
@@ -4344,8 +4344,8 @@ def pivot(self, index=None, columns=None, values=None):
             existing index.
         columns : string or object
             Column name to use to make new frame's columns
-        values : string or object, optional
-            Column name to use for populating new frame's values. If not
+        values : string, object or a list of the previous, optional
+            Column name(s) to use for populating new frame's values. If not
```
Comment (jreback) [inline pandas/core/frame.py]: add (0.23.0) after 'list of the previous'

File: pandas/core/frame.py [hunk 14]
```diff
@@ -4390,6 +4391,19 @@ def pivot(self, index=None, columns=None, values=None):
         one  1   2   3
         two  4   5   6
 
+       >>> df.pivot(index='foo', columns='bar', values=['baz', 'zoo'])
+              A  B  C   A  B  C
+        one   1  2  3   x  y  z
+        two   4  5  6   q  w  t
+
+        >>> df.pivot(index='zoo', columns='foo', values=['bar', 'baz'])
```
Comment (jreback) [inline pandas/core/frame.py]: don't show the same example, rather show one with values = a single value

**Review by jreback — CHANGES_REQUESTED** [review]
you need to merge master and push your changes.

**Comment by ibrahimsharaf** [comment]
Hi @jreback, I've pushed my changes, can you review? thanks!

File: pandas/tests/reshape/test_pivot.py (hunk 12 above)
Comment (jreback) [reply pandas/tests/reshape/test_pivot.py]: yep

File: doc/source/whatsnew/v0.22.0.txt [hunk 15]
```diff
@@ -240,4 +240,4 @@ With conda, use
 
 Note that the inconsistency in the return value for all-*NA* series is still
 there for pandas 0.20.3 and earlier. Avoiding pandas 0.21 will only help with
-the empty case.
```
Comment (jreback) [inline doc/source/whatsnew/v0.22.0.txt]: can you revert this file

File: pandas/core/frame.py [hunk 16]
```diff
@@ -4355,8 +4355,8 @@ def pivot(self, index=None, columns=None, values=None):
             existing index.
         columns : string or object
             Column name to use to make new frame's columns
-        values : string or object, optional
-            Column name to use for populating new frame's values. If not
+        values : string, object or (0.23.0) a list of the previous, optional
```
Comment (jreback) [inline pandas/core/frame.py]: put the version *after* the word list

File: pandas/core/frame.py [hunk 17]
```diff
@@ -4401,6 +4402,15 @@ def pivot(self, index=None, columns=None, values=None):
         one  1   2   3
         two  4   5   6
 
+        >>> df.pivot(index='foo', columns='bar', values=['baz'])
```
Comment (jreback) [inline pandas/core/frame.py]: oh we already have this example, I c, ok you can remove this one (with ``['bar']``)

File: pandas/core/reshape/reshape.py (hunk 10 above)
Comment (jreback) [reply pandas/core/reshape/reshape.py]: did you need to check if its a tuple?

File: pandas/tests/reshape/test_pivot.py [hunk 18]
```diff
@@ -371,6 +371,33 @@ def test_pivot_periods(self):
...
+        pd.Series(['bar', 'baz']), pd.Index(['bar', 'baz'])
+    ])
+    def test_pivot_with_list_like_values(self, values):
+        # issue #17160
+        df = pd.DataFrame({'foo': ['one', 'one', 'one', 'two', 'two', 'two'],
+                           'bar': ['A', 'B', 'C', 'A', 'B', 'C'],
+                           'baz': [1, 2, 3, 4, 5, 6],
+                           'zoo': ['x', 'y', 'z', 'q', 'w', 't']})
+
+        result = df.pivot(index='zoo', columns='foo', values=values)
```
Comment (jreback) [inline pandas/tests/reshape/test_pivot.py]: add a test with values as a tuple (it should fail)

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.22.0.txt (hunk 15 above)
Comment (ibrahimsharaf) [reply doc/source/whatsnew/v0.22.0.txt]: sure

File: pandas/core/reshape/reshape.py (hunk 10 above)
Comment (ibrahimsharaf) [reply pandas/core/reshape/reshape.py]: Yes, I'm trying to figure out why the MultiIndex with a tuple of values test is giving me an error, would be great if you give me a hand!

Comment (ibrahimsharaf) [reply pandas/core/reshape/reshape.py]: here @jreback: https://travis-ci.org/pandas-dev/pandas/jobs/325099021#L2255

**Comment by ibrahimsharaf** [comment]
Hi @jreback, can you give me a hand in here? https://travis-ci.org/pandas-dev/pandas/jobs/325099021#L2255

**Comment by ibrahimsharaf** [comment]
Ping @jreback @jorisvandenbossche, shall we finish this PR?

**Comment by TomAugspurger** [comment]
Updated on top of master, @ibrahimsharaf do you want to check the merge, I'm not sure if I got your changes 100% correct. We'll see if the tests pass.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/reshape/test_pivot.py [hunk 19]
```diff
@@ -371,6 +371,78 @@ def test_pivot_periods(self):
...
+        result = df.pivot(index=('bar', 'first'), columns=('bar', 'second'),
+                          values=('baz', 'first'))
+
+        data = {'A': Series([1, 4], index=['one', 'two']),
+                'B': Series([2, 5], index=['one', 'two']),
+                'C': Series([3, 6], index=['one', 'two'])}
+        expected = DataFrame(data)
+        tm.assert_frame_equal(result, expected)
+
+    @pytest.mark.xfail(reason='tuple is seen as a single column name')
```
[the remark this answers is not in the document]
Comment (ibrahimsharaf) [reply pandas/tests/reshape/test_pivot.py]: Shouldn't this test be failing as we agreed?

[the remark this answers is not in the document]
Comment (jreback) [reply pandas/tests/reshape/test_pivot.py]: if you are asserting that it is failing (I agree it should), then you need to check for the exception

```
with pytest.raises(KeyError):
   ...

```

xfai is for tests which are either not implemented / complete.

File: pandas/core/reshape/reshape.py [hunk 20]
```diff
@@ -392,16 +392,20 @@ def pivot(self, index=None, columns=None, values=None):
...
             index = self.index
         else:
             index = self[index]
-        indexed = self._constructor_sliced(
-            self[values].values,
-            index=MultiIndex.from_arrays([index, self[columns]]))
-        return indexed.unstack(columns)
+        index = MultiIndex.from_arrays([index, self[columns]])
+
+        if is_list_like(values) and not isinstance(values, tuple):
```
Comment (jreback) [inline pandas/core/reshape/reshape.py]: can you add a comment here on why excluding tuples

File: pandas/tests/reshape/test_pivot.py [hunk 21]
```diff
@@ -371,6 +371,64 @@ def test_pivot_periods(self):
...
+
+    def test_pivot_with_tuple_of_values(self):
+        # issue #17160
+        df = pd.DataFrame({'foo': ['one', 'one', 'one', 'two', 'two', 'two'],
+                           'bar': ['A', 'B', 'C', 'A', 'B', 'C'],
+                           'baz': [1, 2, 3, 4, 5, 6],
+                           'zoo': ['x', 'y', 'z', 'q', 'w', 't']})
+        with pytest.raises(KeyError):
+            # tuple is seen as a single column name
+            result = df.pivot(index='zoo', columns='foo', values=('bar', 'baz'))
```
Comment (jreback) [inline pandas/tests/reshape/test_pivot.py]: don;t need the result = here (its a lint error)

**Review by jreback — CHANGES_REQUESTED** [review]
minor comments. looks good otherwise. ping on green.

**Comment by ibrahimsharaf** [comment]
Hi @jreback, thanks for review. We still have a KeyError exception with the MultiIndex failing test, how do you think it should be fixed?

**Comment by jreback** [comment]
can i paste what case is failing

**Comment by ibrahimsharaf** [comment]
Here: https://travis-ci.org/pandas-dev/pandas/jobs/348034399#L2303

**Comment by TomAugspurger** [comment]
https://github.com/pandas-dev/pandas/issues/19966 for the isssue causing that test to fail.

We can see how hard that is to fix. I'd say just xfail the test for now, rather than trying to work around it.

File: pandas/core/frame.py [hunk 22]
```diff
@@ -4375,26 +4375,31 @@ def pivot(self, index=None, columns=None, values=None):
...
-        >>> df.pivot(index='foo', columns='bar')['baz']
+        >>> df.pivot(index='foo', columns='bar', values='baz')
              A   B   C
         one  1   2   3
         two  4   5   6
 
+       >>> df.pivot(index='foo', columns='bar', values=['baz', 'zoo'])
+              A  B  C   A  B  C
+        one   1  2  3   x  y  z
+        two   4  5  6   q  w  t
```
[the remark this answers is not in the document]
Comment (ibrahimsharaf) [reply pandas/core/frame.py]: Yes, that's right. Fixed.

**Comment by ibrahimsharaf** [comment]
Shall you guys take a look? @jreback @jorisvandenbossche @TomAugspurger

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm @ibrahimsharaf 

@TomAugspurger @jorisvandenbossche might have some comments

File: pandas/core/frame.py [hunk 23]
```diff
@@ -4955,9 +4955,9 @@ def pivot(self, index=None, columns=None, values=None):
             Column to use to make new frame's index. If None, uses
             existing index.
         columns : string or object
-            Column to use to make new frame's columns.
-        values : string or object, optional
-            Column to use for populating new frame's values. If not
+            Column name to use to make new frame's columns
```
Comment (jorisvandenbossche) [inline pandas/core/frame.py]: This is probably due to merging master, but you can you undo this change? (there have been changes to this docstring on master, and you have accidentally reverted some of those changes)

File: pandas/core/frame.py [hunk 24]
```diff
@@ -4955,9 +4955,9 @@ def pivot(self, index=None, columns=None, values=None):
...
+        values : string, object or a list (0.23.0) of the previous, optional
+            Column name(s) to use for populating new frame's values. If not
```
Comment (jorisvandenbossche) [inline pandas/core/frame.py]: `Column name(s)` -> `Column(s)` to be consistent with above

File: pandas/core/frame.py [hunk 25]
```diff
@@ -4955,9 +4955,9 @@ def pivot(self, index=None, columns=None, values=None):
             Column to use to make new frame's index. If None, uses
             existing index.
         columns : string or object
-            Column to use to make new frame's columns.
-        values : string or object, optional
-            Column to use for populating new frame's values. If not
+            Column name to use to make new frame's columns
+        values : string, object or a list (0.23.0) of the previous, optional
```
Comment (jorisvandenbossche) [inline pandas/core/frame.py]: Instead of mentioning the `(0.23.0)` here, can you instead add a 

```
.. versionchanged:: 0.23.0
    Also accept list of column names.
```

to the parameter description?

File: pandas/core/frame.py [hunk 26]
```diff
@@ -5011,8 +5012,14 @@ def pivot(self, index=None, columns=None, values=None):
         one  1   2   3
         two  4   5   6
 
-        A ValueError is raised if there are any duplicates.
+        >>> df.pivot(index='foo', columns='bar', values=['baz', 'zoo'])
+                baz       zoo
```
Comment (jorisvandenbossche) [inline pandas/core/frame.py]: The alignment seems a bit off here. I think the 'baz' should align with the 'A' below. Best to check how it is in the console output.

File: pandas/core/frame.py [hunk 27]
```diff
@@ -5022,10 +5029,8 @@ def pivot(self, index=None, columns=None, values=None):
         1  one   A    2
         2  two   B    3
         3  two   C    4
-
         Notice that the first two rows are the same for our `index`
         and `columns` arguments.
-
```
Comment (jorisvandenbossche) [inline pandas/core/frame.py]: Can you also undo this removal of blank lines?

Comment (ibrahimsharaf) [reply pandas/core/frame.py]: All fixed now @jorisvandenbossche.

**Comment by ibrahimsharaf** [comment]
Can you @jreback @jorisvandenbossche restart Travis build?

**Comment by ibrahimsharaf** [comment]
ping @jorisvandenbossche

**Comment by jreback** [comment]
lgtm. @jorisvandenbossche merge when ok.

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
@ibrahimsharaf Thanks a lot!
